### PR TITLE
Add support for Dremio as a data source

### PR DESCRIPTION
**Pull request by mitchdevenport**
Adds support for using a Dremio instance as the backend source for a dataset and as an authentication provider.

While authoring this PR it became evident that any source which uses Flight could generically use the same `datasource`, so I have generalized the existing `spiceai` datasource into a `flight` datasource.

### CODE REVIEW

File: crates/runtime/src/auth.rs [hunk 1]
```diff
@@ -29,7 +30,15 @@ pub trait AuthProvider {
...
-    fn get_token(&self) -> String;
+    fn get_token(&self) -> String {
+        String::new()
+    }
+    fn get_username(&self) -> String {
+        String::new()
+    }
+    fn get_password(&self) -> String {
+        String::new()
+    }
```
Comment (phillipleblanc) [inline crates/runtime/src/auth.rs:41]: These should return Option<String> and return None as the default implementation.

Also - why do we have get_token and get_password? This feels like the abstraction is breaking down. We could still just keep `get_token` if we encoded the username/password for the flight client like `{username}:{password}` right?

File: crates/runtime/src/databackend/memtable.rs [hunk 2]
```diff
@@ -24,7 +24,7 @@ impl MemTableBackend {
     pub fn new(ctx: Arc<SessionContext>, name: &str) -> Self {
         MemTableBackend {
             ctx,
-            name: name.replace('.', "_").to_string(),
+            name: crate::datafusion::get_internal_dataset_name(name),
```
Comment (phillipleblanc) [inline crates/runtime/src/databackend/memtable.rs:27]: With the new model of `from` - I don't think we need to do this anymore

Comment (lukekim) [reply crates/runtime/src/databackend/memtable.rs:27]: We probably need some validation on valid names though (and to spec what is valid). Probably whatever is a valid GitHub repo and org name is a good start.

File: crates/runtime/src/auth.rs (hunk 1 above)
Comment (mitchdevenport) [reply crates/runtime/src/auth.rs:41]: Spoke offline, we will revisit the auth interface

File: crates/runtime/src/databackend/memtable.rs (hunk 2 above)
Comment (mitchdevenport) [reply crates/runtime/src/databackend/memtable.rs:27]: Deferring discussion of this to the discussion on the still open PR: https://github.com/spiceai/spiceai/pull/632#discussion_r1472242352

**Pull request merged**